internal/service/cloudwatch/contributor_managed_insight_rule.go

```diff
@@ -0,0 +1,400 @@
+		)
+		return
+	}
+
+	if out.RuleState != nil && out.RuleState.RuleName != nil {
+		state.RuleName = fwflex.StringValueToFramework(ctx, *out.RuleState.RuleName)
+		cmirARN := r.Meta().RegionalARN(ctx, "cloudwatch", fmt.Sprintf("insight-rule/%s", state.RuleName.ValueString()))
+		state.ARN = fwflex.StringValueToFramework(ctx, cmirARN)
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+}
+
+func (r *resourceContributorManagedInsightRule) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var old, new resourceContributorManagedInsightRuleData
+
+	resp.Diagnostics.Append(req.State.Get(ctx, &old)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &new)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	conn := r.Meta().CloudWatchClient(ctx)
+
+	if !new.State.IsNull() && !old.State.Equal(new.State) {
+		if new.State.ValueString() == "ENABLED" || new.State.IsNull() {
+			_, err := conn.PutManagedInsightRules(ctx, &cloudwatch.PutManagedInsightRulesInput{
+				ManagedRules: []awstypes.ManagedRule{
+					{
+						ResourceARN:  new.ResourceArn.ValueStringPointer(),
+						TemplateName: new.TemplateName.ValueStringPointer(),
+					},
+				},
+			})
+			if err != nil {
+				resp.Diagnostics.AddError(
+					create.ProblemStandardMessage(names.CloudWatch, create.ErrActionUpdating, ResNameContributorManagedInsightRule, new.ResourceArn.String(), err),
+					err.Error(),
+				)
+			}
+		} else if new.State.ValueString() == "DISABLED" {
+			rule, err := findContributorManagedInsightRuleDescriptionByTemplateName(ctx, conn, new.ResourceArn.ValueString(), new.TemplateName.ValueString())
+			if err != nil {
+				resp.Diagnostics.AddError(
+					create.ProblemStandardMessage(names.CloudWatch, create.ErrActionUpdating, ResNameContributorManagedInsightRule, new.ResourceArn.String(), err),
+					err.Error(),
+				)
+			}
+			_, err = conn.DisableInsightRules(ctx, &cloudwatch.DisableInsightRulesInput{
+				RuleNames: []string{*rule.RuleState.RuleName},
+			})
+			if err != nil {
+				resp.Diagnostics.AddError(
+					create.ProblemStandardMessage(names.CloudWatch, create.ErrActionUpdating, ResNameContributorManagedInsightRule, new.ResourceArn.String(), err),
+					err.Error(),
+				)
+			}
+		}
+	}
+
+	new.RuleName = old.RuleName
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &new)...)
+}
+
+func (r *resourceContributorManagedInsightRule) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	conn := r.Meta().CloudWatchClient(ctx)
+
+	var state resourceContributorManagedInsightRuleData
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	rule, err := findContributorManagedInsightRuleDescriptionByTemplateName(ctx, conn, state.ResourceArn.ValueString(), state.TemplateName.ValueString())
+	if err != nil {
+		if errs.IsA[*awstypes.ResourceNotFoundException](err) {
```

Comment:
it shouldn't be possible to get this error message type here if this find function is returning a provider defined `NotFound` error